packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1466 @@
+
+        unchecked {
+            _vault.asset.totalKro += assets;
+            _vault.asset.totalKroShares += shares;
+            _vault.kroDelegators[owner].shares += shares;
+
+            if (owner == validator) {
+                _vault.asset.validatorKro += assets;
+            }
+        }
+
+        _updateWeightTree(validator);
+    }
+
+    /**
+     * @notice Internal function to delegate KGH to the validator.
+     *
+     * @param validator Address of the validator.
+     * @param tokenId   Token Id of the KGH.
+     * @param kroInKgh  The amount of KRO in the KGH.
+     * @param kroShares The amount of KRO shares to receive.
+     * @param kghShares The amount of KGH shares to receive.
+     */
+    function _delegateKgh(
+        address validator,
+        uint256 tokenId,
+        uint128 kroInKgh,
+        uint128 kroShares,
+        uint128 kghShares
+    ) internal virtual {
+        Vault storage _vault = vault[validator];
+        IERC721(KGH).safeTransferFrom(msg.sender, address(this), tokenId);
+
+        unchecked {
+            _vault.asset.totalKro += kroInKgh;
+            _vault.asset.totalKroInKgh += kroInKgh;
+            _vault.asset.totalKgh += 1;
+
+            _vault.asset.totalKroShares += kroShares;
+            _vault.asset.totalKghShares += kghShares;
+            _vault.kghDelegators[msg.sender].kroShares += kroShares;
+            _vault.kghDelegators[msg.sender].kghShares[tokenId] += kghShares;
+        }
+
+        _updateWeightTree(validator);
+    }
+
+    /**
+     * @notice Internal function to delegate KGHs to the validator.
+     *
+     * @param validator Address of the validator.
+     * @param kghCount  The number of KGHs to delegate.
+     * @param kroInKghs The amount of KRO in the KGHs.
+     * @param kroShares The amount of KRO shares to receive.
+     * @param kghShares The amount of KGH shares to receive.
+     */
+    function _delegateKghBatch(
+        address validator,
+        uint128 kghCount,
+        uint128 kroInKghs,
+        uint128 kroShares,
+        uint128 kghShares
+    ) internal virtual {
+        Vault storage _vault = vault[validator];
+
+        unchecked {
+            _vault.asset.totalKro += kroInKghs;
+            _vault.asset.totalKroInKgh += kroInKghs;
+            _vault.asset.totalKgh += kghCount;
+
+            _vault.asset.totalKroShares += kroShares;
+            _vault.asset.totalKghShares += kghShares;
+            _vault.kghDelegators[msg.sender].kroShares += kroShares;
+        }
+
+        _updateWeightTree(validator);
+    }
+
+    /**
+     * @notice Internal function to undelegate KRO from the validator.
+     *
+     * @param validator Address of the validator.
+     * @param owner     Address of the delegator.
+     * @param assets    The amount of KRO to undelegate.
+     * @param shares    The amount of shares to undelegate.
+     */
+    function _initUndelegate(
```

Comment:
I think if validator self-undelegates and leaves less than `MIN_REGISTER_AMOUNT`, he should remove himself from the weight tree. So in this function, the logic to check it and remove from tree should be added. wdyt?

Reply:
or more strictly, if less than `MIN_START_AMOUNT`?

Reply:
But if then, there could be an edge case that delegators withdraw all their tokens from the vault, so the validator vault has low amount of tokens but still that validator remains at the tree. How do you think about this?

Reply:
Yes, such edge cases can be problematic, but I'm worried that undelegation will be too costly for delegators of small amount. It also seems to happen fairly often that the total KRO goes below `MIN_START_AMOUNT` by an undelegation, but it doesn't seem to happen as often that all delegators leave but validators keep their KRO above `MIN_REGISTER_AMOUNT`.

Reply:
A validator should not get an opportunity to submit output if he has less amount of KRO than `minStartAmount`, not `minRegisterAmount`? I agree with the point that the gas that users spend will increase in some cases, but I think it will be somewhat problematic if unqualified validators are not removed from the tree.

Reply:
Yes the validator misses the output submission opportunity when he has less than `minStartAmount`. I agree with your point, then let's just remain it and discuss about it after investigating the gas usage when we do QA.